transport/transport-runtime/src/main/java/com/google/android/datatransport/runtime/scheduling/persistence/SQLiteEventStore.java

```diff
@@ -530,6 +537,145 @@ private static Encoding toEncoding(@Nullable String value) {
     } while (true);
   }
 
+  /**
+   * Record log event dropped.
+   *
+   * @param eventsDroppedCount number of events dropped
+   * @param reason reason why events are dropped
+   * @param logSource log source of the dropped events
+   */
+  @Override
+  public void recordLogEventDropped(
+      long eventsDroppedCount, LogEventDropped.Reason reason, String logSource) {
+    if (!isLogEventDroppedExist(reason, logSource)) {
+      ContentValues metrics = new ContentValues();
+      metrics.put("log_source", logSource);
+      metrics.put("reason", reason.getNumber());
+      metrics.put("events_dropped_count", eventsDroppedCount);
+      inTransaction(
+          db -> {
+            db.insert("log_event_dropped", null, metrics);
+            return null;
+          });
+    } else {
+      String query =
+          "UPDATE log_event_dropped SET events_dropped_count = events_dropped_count + "
+              + eventsDroppedCount
+              + " WHERE log_source = ? AND reason = ?";
+      inTransaction(
+          db -> {
+            db.execSQL(query, new String[] {logSource, Integer.toString(reason.getNumber())});
+            return null;
+          });
+    }
+  }
+
+  private boolean isLogEventDroppedExist(LogEventDropped.Reason reason, String logSource) {
+    String query = "SELECT 1 FROM log_event_dropped" + " WHERE log_source = ? AND reason = ?";
+    String[] selectionArgs = new String[] {logSource, Integer.toString(reason.getNumber())};
+    return inTransaction(
+        db -> tryWithCursor(db.rawQuery(query, selectionArgs), cursor -> cursor.getCount() > 0));
+  }
+
+  /**
+   * Load full-populated ClientMetrics.
+   *
+   * @return the returned ClientMetrics is fully populated with persisted and non-persisted fields.
+   */
+  @Override
+  public ClientMetrics loadClientMetrics() {
+    long currentTime = wallClock.getTime();
+    ClientMetrics.Builder clientMetricsBuilder = ClientMetrics.newBuilder();
+    Map<String, List<LogEventDropped>> metricsMap = new HashMap<>();
+    String query = "SELECT log_source, reason, events_dropped_count FROM log_event_dropped";
+
+    return inTransaction(
+        db ->
+            tryWithCursor(
+                db.rawQuery(query, new String[] {}),
+                cursor -> {
+                  while (cursor.moveToNext()) {
+                    String logSource = cursor.getString(0);
+                    LogEventDropped.Reason reason =
+                        LogEventDropped.Reason.values()[cursor.getInt(1)];
```

Comment:
Using `values()` is not accurate as enum value indexes are not the same as the protobuf "number"(I mean it happens to work now but is fragile).

Looks like codegen is missing this functionality, created b/192450103.

We should either fix b/192450103 or come up with a workaround by e.g. creating a mapping from ints to Reason values.

Reply:
having that functionality would be great, just did some research on `values()` as this [doc](https://docs.oracle.com/javase/specs/jls/se7/html/jls-8.html#jls-8.9.2) states: `values()` returns an array containing the constants of this enum type, **in the order** they're declared. And the firebase encoder generates those enum in order and the number of each enum is assigned from 0, plus the proto is also ordered from 0. In that case, this could be accurate. since they are correct, shall we keep it in that way and come back to change it until b/192450103 solved?

Reply:
The issue is that nothing is preventing me from adding values to the proto out of order, e.g.

```proto
enum Foo {
  UNKNOWN = 0;
  BAR = 2;
  BAZ = 1;
}
```

This is why I would not recommend relying on the order which just `happens` to be identical to the order of java enums.

Because it is fragile, I suggest we choose the `ProtoUtil` way that you described or even just a private static method inside this class with a `switch(number)`, since this is the only place that needs this mapping

Reply:
sg, will do that